modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java

```diff
@@ -110,6 +110,10 @@ protected boolean initializeCamera() {
             if (mCameraID != null) {
                 Log.i(LOGTAG, "Opening camera: " + mCameraID);
                 manager.openCamera(mCameraID, mStateCallback, mBackgroundHandler);
+            } else { // make JavaCamera2View's logic consistent with JavaCameraView's.
+                Log.i(LOGTAG, "Trying to open camera with new open(" + mCameraIndex + ")");
+                mCameraID = Integer.toString(mCameraIndex);
```

Comment:
I believe it should be:
```
... check range (mCameraIndex) or lets exception be triggered ...
mCameraID = camList[mCameraIndex];
```

Reply:
I agree that there should be error-checking. To be consistent with `JavaCameraView`, I think we should check `if (Arrays.asList(camList).contains(Integer.toString(mCameraIndex)))`. How do you think about the following code?

```suggestion
Log.i(LOGTAG, "Trying to open camera with new open(" + mCameraIndex + ")");
mCameraID = Integer.toString(mCameraIndex);
if (Arrays.asList(camList).contains(mCameraID)) {
    manager.openCamera(mCameraID, mStateCallback, mBackgroundHandler);
} else {
    // CAMERA_DISCONNECTED is used when the camera id is no longer valid
    throw new CameraAccessException(CameraAccessException.CAMERA_DISCONNECTED);
}
```

Reply:
> mCameraIndex

is "int", but

> mCameraID

is string. "Integers" string are not expected in mCameraID.

This is wrong in all meanings:

> Integer.toString(mCameraIndex)


Take a look on line 96.

Reply:
As you said, `mCameraIndex` is `int` and `mCameraID` is `String`. For example, `mCameraIndex` can be `0` while `mCameraID` can be `"0"`. `mCameraID` is a stringyfied version of camera index starting from 0. That's why I used `Integer.toString()` to convert `int` to `String`, for example `0` to `"0"`. Although it is used less often, `Integer.toString()` is an equivalent of `String.valueOf()` so that it can also be used for this purpose.

Reply:
Again, it should be:
```
mCameraID = camList[mCameraIndex];
```

Reply:
I think this code is what you want:
```java
Log.i(LOGTAG, "Trying to open camera with the value (" + mCameraIndex + ")");
if (mCameraIndex < camList.length) {
    mCameraID = camList[mCameraIndex];
    manager.openCamera(mCameraID, mStateCallback, mBackgroundHandler);
} else {
    // CAMERA_DISCONNECTED is used when the camera id is no longer valid
    throw new CameraAccessException(CameraAccessException.CAMERA_DISCONNECTED);
}
```

Consider the following situation:
* There are 3 cameras: `camera-0`, `camera-1` and `camera-2`
* A problem occurred on`camera-0`, so that `camList == {"1", "2"}`
* A user manually set `setCameraIndex(1)`
* Then `JavaCameraView` will open `camera-1`; My previous commit will also open `camera-1`; but this new code will open `camList[1]`, which is `camera-2`

I hope this explanation to be clear. If you still want this new code, please tell me. I will commit the new code and squash the previous commits.

Reply:
Right!

Java and C++ uses zero-based indexes. So,
- `0` opens the first available camera (`camList[0]`)
- `1` opens the second camera (`camList[1]`)
- etc

We should not assume that `camList == {"1", "2"}` or contains direct integers.

> setCameraIndex(1)
> will open `camera-1`

No, it should open the second available camera (second from **the list of available cameras**).

Reply:
I committed the new code and squashed all previous commits. After the automated build and checks it will become available. Thanks for your hard work on OpenCV! Many people around the world, including me, are benefiting from you.